Jenkinsfile

```diff
@@ -4,122 +4,96 @@ pipeline {
+                            }
+                        }
+                        stage('Test') {
+                            steps {
+                                sh "make subsystem-run"
+                            }
+                        }
+                    }
+                }
 
-        stage('Deploy for subsystem') {
-            steps {
-                withCredentials([usernamePassword(credentialsId: 'dockerio_cred', passwordVariable: 'PASS', usernameVariable: 'USER')]) {
-                    sh '''docker login docker.io -u $USER -p $PASS'''
+                stage("OnPrem") {
+                    agent { label 'centos_worker' }
+                    stages {
+                        stage("Build") {
+                            steps {
+                                sh "make build-onprem"
+                                sh "make deploy-onprem"
+                            }
+                        }
+                        stage("Test") {
+                            steps {
+
+                                sh "make test-onprem"
+                            }
+                        }
+                    }
                 }
-                sh '''export PATH=$PATH:/usr/local/go/bin; make jenkins-deploy-for-subsystem'''
-                sh '''# Dump pod statuses;kubectl get pods -A'''
             }
         }
 
-        stage('Subsystem-test') {
-            steps {
-                sh '''export PATH=$PATH:/usr/local/go/bin;make subsystem-run'''
-            }
-        }
 
-        stage('clear deployment after subsystem test') {
-            steps {
-                sh 'make clear-deployment'
-            }
-        }
 
-        stage('publish images on push to master') {
-            when {
-                branch 'master'
-            }
 
+        stage('Publish') {
+            when { branch 'master'}
             steps {
-                withCredentials([usernamePassword(credentialsId: 'ocpmetal_cred', passwordVariable: 'PASS', usernameVariable: 'USER')]) {
-                    sh '''docker login quay.io -u $USER -p $PASS'''
-                }
-
-                sh '''docker tag  ${SERVICE} quay.io/ocpmetal/assisted-service:latest'''
-                sh '''docker tag  ${SERVICE} quay.io/ocpmetal/assisted-service:${GIT_COMMIT}'''
-                sh '''docker push quay.io/ocpmetal/assisted-service:latest'''
-                sh '''docker push quay.io/ocpmetal/assisted-service:${GIT_COMMIT}'''
-
-                sh '''docker tag  ${ISO_CREATION} quay.io/ocpmetal/assisted-iso-create:latest'''
-                sh '''docker tag  ${ISO_CREATION} quay.io/ocpmetal/assisted-iso-create:${GIT_COMMIT}'''
-                sh '''docker push quay.io/ocpmetal/assisted-iso-create:latest'''
-                sh '''docker push quay.io/ocpmetal/assisted-iso-create:${GIT_COMMIT}'''
+                sh "docker login quay.io -u ${QUAY_IO_CREDS_USR} -p ${QUAY_IO_CREDS_PSW}"
+                sh "make publish"
```

Comment:
Is there a guaranty that it will run on the host that build the images? 
You run the build on parallel so it this part can run on a host that execute the onprem?

Reply:
Everything under "agent" runs on the same host unless there's a new "agent" section

Reply:
So to run something in parallel you need to define parallel and add a different agent to each step?

Reply:
When you define a `parallel` section it would fork to multiple threads.
For each stage which defines an `agent` it looks for another `agent.
So in this case - opening a `parallel` section having two stages - only one with another agent - will run the 1st one on a new thread, and the 2nd thread will call another node and waits for its termination.